include/seqan3/container/detail/ra_iterator.hpp

```diff
@@ -0,0 +1,216 @@
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+// Authors: Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+// ============================================================================
+// Implementation of the (const) iterator given a container reference.
+// ============================================================================
+
+#pragma once
+
+#include <iomanip>
+#include <limits>
+#include <type_traits>
+
+#include <range/v3/all.hpp>
+#include <seqan3/container/concepts.hpp>
+
+namespace seqan3::detail
+{
+
+template <typename container_type, bool is_const=false>
+    requires random_access_range_concept<container_type> && sized_range_concept<container_type>
+class ra_iterator
+{
+
+private:
+    // by default iterator is end iterator
+    container_type & host{*(container_type*)0};
+    using position_type = uint8_t;
+    position_type pos{std::numeric_limits<uint8_t>::max()};
+public:
+    using difference_type = int8_t;
+    using value_type = typename container_type::value_type;
+    using reference = std::conditional_t<is_const, value_type const &, value_type &>;
+    using const_reference = typename container_type::const_reference;
+    using pointer = std::conditional_t<is_const, value_type const *, value_type *>;
+    using category_type = std::random_access_iterator_tag;
+
+    ra_iterator() {}
+
+    ra_iterator(container_type & host, bool const _at_end = false) : host{host}, pos{0}
+    {
+        if (_at_end)
+            pos = std::numeric_limits<difference_type>::max();
+    }
+
+    // copy constructor
+    ra_iterator(ra_iterator const & _it) : host{_it.host}, pos{_it.pos} {}
+
+    // copy via assignment
+    ra_iterator& operator=(ra_iterator const & rhs)
+    {
+        pos = rhs.pos;
+        return *this;
+    }
+
+    ~ra_iterator() = default;
+
+    // return element currently pointed at
+    reference operator*() const
```

Comment:
`reference` is "generic" and resolved either to `value_type` or `const value_type`. Defined top: `using reference = std::conditional_t<is_const, value_type const &, value_type &>;`. We could later use your metafunctions to return the non-const or const `value_type` of the container.